lib/gst/player/gstplayer.c

```diff
@@ -1121,6 +1167,464 @@ element_cb (GstBus * bus, GstMessage * msg, gpointer user_data)
   }
 }
 
+static void
+player_set_flag (GstPlayer *self, gint pos)
+{
+  gint flags;
+
+  g_object_get (self->priv->playbin, "flags", &flags, NULL);
+  flags |= pos;
+  g_object_set (self->priv->playbin, "flags", &flags, NULL);
+}
+
+static void
+player_clear_flag (GstPlayer *self, gint pos)
+{
+  gint flags;
+
+  g_object_get (self->priv->playbin, "flags", &flags, NULL);
+  flags &= ~pos;
+  g_object_set (self->priv->playbin, "flags", &flags, NULL);
+}
+
+typedef struct
+{
+  GstPlayer *player;
+  GstPlayerMediaInfo  *info;
+  guint signal;
+}MediaInfoUpdatedSignalData;
+
+static gboolean
+media_info_updated_dispatch (gpointer user_data)
+{
+  MediaInfoUpdatedSignalData  *data = user_data;
+
+  g_signal_emit (data->player, signals[data->signal], 0, data->info);
+
+  return FALSE;
+}
+
+static void
+free_media_info_updated_signal_data (MediaInfoUpdatedSignalData *data)
+{
+  g_slice_free (MediaInfoUpdatedSignalData, data);
+}
+
+static void
+emit_media_updated_signal (GstPlayer *self, GstPlayerMediaInfo *info,
+  guint signal)
+{
+  if (self->priv->dispatch_to_main_context) {
+    MediaInfoUpdatedSignalData *data = g_slice_new
+                                        (MediaInfoUpdatedSignalData);
+
+    data->player = self;
+    g_mutex_lock (&self->priv->lock);
+    if (self->priv->media_info)
+      g_object_unref (self->priv->media_info);
+    self->priv->media_info = info;
+    data->info   = info;
+    g_mutex_unlock (&self->priv->lock);
+    data->signal = signal;
+
+    g_main_context_invoke_full (self->priv->application_context,
+                    G_PRIORITY_DEFAULT,
+                    media_info_updated_dispatch,
+                    data,
+                    (GDestroyNotify) free_media_info_updated_signal_data);
+  } else {
+    g_signal_emit (self, signals[signal], 0);
```

Comment:
Okay maybe I am missing something or have some confusion.

* self->priv->media_info is pointing to internal media_info. this media_info is constantly been updated by *-tags-changed or *-changed callbacks.
* when *-tags-changed or *-changed callback is executed then we update the content of self->priv->media_info and create a new copy from the current media_info and emit the newly created copy to user. In this way, the media_info received by user app is immutable
so our sequence is:
lock()
new_media_info = gst_player_media_info_copy (self->priv->media_info);
unlock()
emit (new_media_info)
unref(new_meda_info)

Do you see anything wrong in this approach ?

If we give reference of self->priv->media_info then its not immutable and it can change while user app is still inside the emit signal callback.